Fix fitError in Permit plugin not handled perfectly

kerthcet opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind bug
/sig scheduling

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

We used to record the failed plugin name in PermitPlugin, but we didn't handle it correctly in eventhandler.go for we only register the `UnschedulablePlugins` to podInfo iff we hit a fitError, see https://github.com/kubernetes/kubernetes/blob/b1d43aa1ddbca80f75c75a8dfe67ba9d40272384/pkg/scheduler/schedule_one.go#L928-L930

So in this PR, when we failed in Permit plugin, we'll return a fitError inside the status.

Besides this, we'll add another support for reserve plugin, tracked by https://github.com/kubernetes/kubernetes/pull/118529

cc @Huang-Wei @alculquicondor  @sanposhiho 

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
None
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

kerthcet commented:
/hold 
for discussion

Huang-Wei commented:
In short term, I'm fine with returning a `FitError` upon failures in Permit(). In long run, we can come up with a better struct with explicit semantics to replace `FitError` - the current type assertion looks a bit ugly TBH..

https://github.com/kubernetes/kubernetes/blob/d9714078f821ccae1e2d50ea69c081d73a255f1f/pkg/scheduler/schedule_one.go#L141

sanposhiho commented on pkg/scheduler/framework/runtime/framework.go:
```diff
@@ -1332,9 +1332,17 @@ func (f *frameworkImpl) RunPermitPlugins(ctx context.Context, state *framework.C
+					NumAllNodes: 1,
+					Pod:         pod,
+					Diagnosis: framework.Diagnosis{
+						PreFilterMsg:         msg,
```
PermitMsg?

kerthcet replied:
Changed the logic, PTAL.

kerthcet commented:
/retest

kerthcet commented:
> In short term, I'm fine with returning a `FitError` upon failures in Permit(). In long run, we can come up with a better struct with explicit semantics to replace `FitError` - the current type assertion looks a bit ugly TBH..
> 
> https://github.com/kubernetes/kubernetes/blob/d9714078f821ccae1e2d50ea69c081d73a255f1f/pkg/scheduler/schedule_one.go#L141

Yes, that needs a deep refactor, let me focus on the work after we merged these two PRs.

alculquicondor commented:
/assign @ahg-g

ahg-g commented on test/integration/scheduler/rescheduling_test.go:
```diff
@@ -0,0 +1,172 @@
+		// firstSchedulingError means for the first time in scheduling,
+		// pod encounters error, otherwise, pod is unschedulable.
+		// By default, pod is unschedulable for the first time.
+		firstSchedulingError bool
```
this indicates an expected behavior not an "action", so lets prefix it with "want" like the other two below.

kerthcet replied:
Changed to wantFirstSchedulingError

ahg-g commented on test/integration/scheduler/rescheduling_test.go:
```diff
@@ -0,0 +1,172 @@
+type PermitPlugin struct {
+	name            string
+	statusCode      framework.Code
+	numPermitCalled int
```
do you need this?

kerthcet replied:
Yes, it's designed as
- For the first time, it will be scheduled in error/unschedulable
- For the second time, it will be scheduled successfully（by default)

So numPermitCalled helps to control this.

kerthcet commented on test/integration/scheduler/util.go:
```diff
@@ -17,21 +17,144 @@ limitations under the License.
-	waitForNodesInCache        = testutils.WaitForNodesInCache
-	waitForPodUnschedulable    = testutils.WaitForPodUnschedulable
-	waitForReflection          = testutils.WaitForReflection
+	createNode                   = testutils.CreateNode
```
I think we can remove them directly, but I'd like to leave it as a follow up, because it touches a lot of codes, make sense?

ahg-g replied:
Then lets do that for this file at least since it is new code added in this PR.

kerthcet replied:
Finished.

kerthcet commented:
ping @ahg-g

ahg-g commented:
pls squash

kerthcet committed:
Fix: fiterror in permit plugin not handled perfectly

We only added failed plulgins, but actually this will not work unless
we make the status with a fitError because we only copy the failured plugins
to podInfo if it is a fitError

Signed-off-by: kerthcet <kerthcet@gmail.com>

kerthcet commented:
> pls squash

Done, thanks for the quick reply.

ahg-g commented:
/lgtm

kerthcet commented:
/hold cancel
